pkg/didcomm/crypto/legacy/authcrypt/encrypt.go

```diff
@@ -0,0 +1,219 @@
+	}
+
+	protectedBytes, err := c.buildProtected(recipients)
+	if err != nil {
+		return nil, err
+	}
+
+	pld := payload
+
+	AAD := base64.URLEncoding.EncodeToString(protectedBytes)
+	print("AAD: ", AAD, "\n")
+
+	// 	Additional data is b64encode(jsonenc(encodedRecipients))
+	// 	see https://github.com/hyperledger/aries-cloudagent-python/blob/master/aries_cloudagent/wallet/crypto.py#L425
+	symPld := chachaCipher.Seal(nil, nonce, pld, []byte(AAD))
+
+	// symPld has a length of len(pld) + poly1035.TagSize
+	// fetch the tag from the tail
+	tag := symPld[len(symPld)-poly1305.TagSize:]
+	// fetch the cipherText from the head (0:up to the trailing tag)
+	cipherText := symPld[0 : len(symPld)-poly1305.TagSize]
+
+	out, err := c.buildJSON(protectedBytes, nonce, cipherText, tag)
+	if err != nil {
+		return nil, err
+	}
+
+	return out, nil
+}
+
+// TODO refactor dupe
+// TODO legacy encryption uses C20P only, enforce
+// createCipher will create and return a new Chacha20Poly1035 cipher for the given nonceSize and symmetric key
+func createCipher(nonceSize int, symKey []byte) (cipher.AEAD, error) {
+	switch nonceSize {
+	case chacha20poly1305.NonceSize:
+		return chacha20poly1305.New(symKey)
+	case chacha20poly1305.NonceSizeX:
+		return chacha20poly1305.NewX(symKey)
+	default:
+		return nil, errors.New("cipher cannot be created with bad nonce size and shared symmetric Key combo")
+	}
+}
+
+// buildJSON builds the JSON message following the legacy format
+func (c *Crypter) buildJSON(protectedBytes, nonce, cipherText, tag []byte) ([]byte, error) {
+
+	envelope := &Envelope{
+		Protected:  base64.URLEncoding.EncodeToString(protectedBytes),
+		IV:         base64.URLEncoding.EncodeToString(nonce),
+		CipherText: base64.URLEncoding.EncodeToString(cipherText),
+		Tag:        base64.URLEncoding.EncodeToString(tag),
+	}
+
+	jsonOut, err := json.Marshal(envelope)
+	if err != nil {
+		return nil, err
+	}
+
+	return jsonOut, nil
+}
+
+func (c *Crypter) buildProtected(recipients []Recipient) ([]byte, error) {
+
+	var enc string
+	switch c.alg {
+	case C20P:
+		enc = "chacha20poly1305_ietf"
+	case XC20P:
+		enc = "xchacha20poly1305_ietf"
+	}
+
+	protected := Protected{
+		Enc:        enc,
+		Typ:        "JWM/1.0",
+		Alg:        "Authcrypt",
+		Recipients: recipients,
+	}
+
+	protectedBytes, err := json.Marshal(protected)
+	if err != nil {
+		return nil, err
+	}
+
+	m, err := prettyPrint(protectedBytes)
+	if err != nil {
+		return nil, err
+	}
+	print(m, "\n")
+
+	return protectedBytes, nil
+}
+
+func (c *Crypter) buildRecipients(cek *[chacha20poly1305.KeySize]byte) ([]Recipient, error) {
+	var encodedRecipients []Recipient
+
+	for _, recKey := range c.recipients {
+		recipient, err := c.buildRecipient(cek, recKey)
+		if err != nil {
+			return nil, err
+		}
+
+		encodedRecipients = append(encodedRecipients, *recipient)
+	}
+
+	return encodedRecipients, nil
+}
+
+// Note: the spec is incorrect with the parameters for encrypted_key: the CEK is the message to encrypt
```

Comment:
Did you raise an issue on aries-rfcs?

Reply:
The spec in question is being amended to a JWE format (that @Baha-sk is working on) anyway - perhaps we should request it be split to keep a spec for the legacy format for backwards-compatibility, and amend that.